File: Cpp/K02.cpp
#include "Golem.h"
#include "K02.h"


K02::K02() = default;


K02::~K02() = default;

#include <iostream>


void K02::question1() {
    constexpr auto seisuu = 3;
    constexpr auto jissuu = 2.6f;
    constexpr auto moji = 'A';

    std::cout << "変数seisuuの値は" << seisuu << "\n";
    std::cout << "変数jissuuの値は" << jissuu << "\n";
    std::cout << "変数mojiの値は" << moji << "\n";
}

void K02::question2() {
    try {
        std::cout << "一つ目の整数は？";
        auto number1 = 0;
        std::cin >> number1;
        std::cout << "二つ目の整数は？";
        auto number2 = 0;
        std::cin >> number2;

        std::cout << number1 << "÷" << number2 << "="
                  << number1 / number2 << "..." << number1 % number2;
    }
    catch (const std::exception &ex) {
        std::cout << ex.what();
    }

}

void K02::question3() {
    try {
        std::cout << "一つ目の商品の値段は？";
        auto priceA = 0;
        std::cin >> priceA;
        std::cout << "個数は？";
        auto amountA = 0;
        std::cin >> amountA;
        std::cout << "二つ目の商品の値段は？";
        auto priceB = 0;
        std::cin >> priceB;
        std::cout << "個数は？";
        auto amountB = 0;
        std::cin >> amountB;

        const auto total = static_cast<float>(priceA * amountA + priceB * amountB) * 1.1f;
        std::cout << "お支払いは税込み￥" << total << "です。";
    }
    catch (const std::exception &ex) {
        std::cout << ex.what();
    }
}

void K02::question4() {
    try {
        Golem golem{};
        std::cout << "ゴーレム　（HP：" << golem.hp << "　防御力：" << golem.defense << "）\n";
        std::cout << "HP：" << golem.hp << "\n";
        std::cout << "今回の攻撃の値を入力してください＞";
        auto damage = 0;
        std::cin >> damage;

        damage = damage - golem.defense > 0 ? damage - golem.defense : 0;

        std::cout << "ダメージは" << damage << "です。\n";
        golem.hp -= damage;
        std::cout << "残りのHPは" << golem.hp << "です。\n";
    }
    catch (const std::exception &ex) {
        std::cout << ex.what();
    }
}

File: Nodejs/K02.ts
import { Golem } from './Golem';
import { IExecutable } from "./IExecutable";
import * as readline from 'readline';
import { format } from 'util';

export class K02 extends IExecutable {
    Question1() {
        let seisuu = 3;
        let jissuu = 2.6;
        let moji = 'A';

        console.log("整数seisuuの値は" + seisuu);
        console.log("実数jissuuの値は" + jissuu);
        console.log("文字mojiの値は" + moji);
        return new Promise<string>((resolve, reject) => {
            resolve();
        });
    }

    Question2(rl?: readline.ReadLine) {
        try {
            let firstNumber: number;
            let secondNumber: number;

            let question1 = () => {
                return new Promise<string>((resolve, reject) => {
                    rl.question("一つ目の整数は？", (value) => {
                        firstNumber = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let question2 = () => {
                return new Promise<string>((resolve, reject) => {
                    rl.question("二つ目の整数は？", (value) => {
                        secondNumber = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let showResult = () => {
                return new Promise<string>(async (resolve, reject) => {
                    await question1();
                    await question2();
                    console.log(firstNumber + "÷" + secondNumber + "="
                        + Math.trunc(firstNumber / secondNumber) + "..." + firstNumber % secondNumber);
                    rl.close();
                    resolve();
                });
            };

            return new Promise<string>(async (resolve, reject) => {
                await showResult();
                resolve();
            });
        } catch (e) {
            console.error(e);
        }
    }

    Question3(rl: readline.ReadLine) {
        try {
            let priceA = 0;
            let amountA = 0;
            let priceB = 0;
            let amountB = 0;

            let question1 = () => {
                return new Promise<string>((resolve, reject) => {
                    rl.question("一つ目の商品の値段は？", (value) => {
                        priceA = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let question2 = () => {
                return new Promise<string>((resolve, reject) => {
                    rl.question("個数は？", (value) => {
                        amountA = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let question3 = () => {
                return new Promise<string>((resolve, reject) => {
                    rl.question("二つ目の商品の値段は？", (value) => {
                        priceB = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let question4 = () => {
                return new Promise<string>((resolve, reject) => {
                    rl.question("個数は？", (value) => {
                        amountB = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let showResult = () => {
                return new Promise(async (resolve, reject) => {
                    await question1();
                    await question2();
                    await question3();
                    await question4();
                    let total = (priceA * amountA + priceB * amountB) * 1.08;
                    console.log("お支払いは税込み￥" + total + "です。");
                    resolve();
                });
            };

            return new Promise<string>(async (resolve, reject) => {
                await showResult();
                resolve();
            });
        } catch (e) {
            console.error(e);
        }
    }

    Question4(rl: readline.ReadLine) {
        try {
            let golem = new Golem();

            console.log(format("ゴーレム　（HP：%d　防御力：%d）", golem.hp, golem.defense));
            console.log("HP：" + golem.hp);

            let damage = 0;

            let question1 = () => {
                return new Promise<string>((resolve, reject) => {
                    rl.question("今回の攻撃の値を入力してください＞", (value) => {
                        damage = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let showResult = () => {
                return new Promise(async (resolve, reject) => {
                    await question1();
                    damage = damage - golem.defense > 0 ? damage - golem.defense : 0;

                    console.log("ダメージは" + damage + "です。");
                    golem.hp -= damage;

                    console.log("残りのHPは" + golem.hp + "です。");
                    resolve();
                });
            }

            return new Promise<string>(async (resolve, reject) => {
                await showResult();
                resolve();
            });
        } catch (e) {
            console.error(e);
        }
    }
}